src/theory/uf/theory_uf.cpp

```diff
@@ -586,6 +653,180 @@ void TheoryUF::eqNotifyDisequal(TNode t1, TNode t2, TNode reason) {
+  std::map< TypeNode, std::vector< Node > > func_eqcs;
+
+  eq::EqClassesIterator eqcs_i = eq::EqClassesIterator( &d_equalityEngine );
+  while( !eqcs_i.isFinished() ){
+    Node eqc = (*eqcs_i);
+    TypeNode tn = eqc.getType();
+    if( tn.isFunction() ){
+      func_eqcs[tn].push_back( eqc );
+      Trace("uf-ho-debug") << "  func eqc : " << tn << " : " << eqc << std::endl;
+    }
+    ++eqcs_i;
+  }
+  
+  for( std::map< TypeNode, std::vector< Node > >::iterator itf = func_eqcs.begin(); 
+       itf != func_eqcs.end(); ++itf ){
+    for( unsigned j=0; j<itf->second.size(); j++ ){
+      for( unsigned k=(j+1); k<itf->second.size(); k++ ){ 
+        // if these equivalence classes are not explicitly disequal, do extensionality to ensure distinctness
+        if( !d_equalityEngine.areDisequal( itf->second[j], itf->second[k], false ) ){
+          Node deq = Rewriter::rewrite( itf->second[j].eqNode( itf->second[k] ).negate() );
+          num_lemmas += applyExtensionality( deq );
+        }
+      }   
+    }
+  }
+  return num_lemmas;
+}
+
+unsigned TheoryUF::applyAppCompletion( TNode n ) {
+  Assert( n.getKind()==kind::APPLY_UF );
+
+  //must expand into APPLY_HO version if not there already
+  Node ret = TheoryUfRewriter::getHoApplyForApplyUf( n );
+  if( !d_equalityEngine.hasTerm( ret ) || !d_equalityEngine.areEqual( ret, n ) ){
+    Node eq = ret.eqNode( n );
+    Trace("uf-ho-lemma") << "uf-ho-lemma : infer, by apply-expand : " << eq << std::endl;
+    d_equalityEngine.assertEquality(eq, true, d_true);
+    return 1;
+  }else{
+    Trace("uf-ho-debug") << "    ...already have " << ret << " == " << n << "." << std::endl;
+    return 0;
+  }
+}
+
+unsigned TheoryUF::checkAppCompletion() {
+  Trace("uf-ho") << "TheoryUF::checkApplyCompletion..." << std::endl;
+  // compute the operators that are relevant (those for which an HO_APPLY exist)
+  std::set< TNode > rlvOp;
+  eq::EqClassesIterator eqcs_i = eq::EqClassesIterator( &d_equalityEngine );
+  std::map< TNode, std::vector< Node > > apply_uf;
+  while( !eqcs_i.isFinished() ){
+    Node eqc = (*eqcs_i);
+    Trace("uf-ho-debug") << "  apply completion : visit eqc " << eqc << std::endl;
+    eq::EqClassIterator eqc_i = eq::EqClassIterator( eqc, &d_equalityEngine );
+    while( !eqc_i.isFinished() ){
+      Node n = *eqc_i;
+      if( n.getKind()==kind::APPLY_UF || n.getKind()==kind::HO_APPLY ){
+        int curr_sum = 0;
+        std::map< TNode, bool > curr_rops;
+        if( n.getKind()==kind::APPLY_UF ){
+          TNode rop = d_equalityEngine.getRepresentative( n.getOperator() );
+          if( rlvOp.find( rop )!=rlvOp.end() ){
+            // try if its operator is relevant
+            curr_sum = applyAppCompletion( n );
+            if( curr_sum>0 ){
+              return curr_sum;
+            }
+          }else{
+            // add to pending list
+            apply_uf[rop].push_back( n );
+          }
+          //arguments are also relevant operators  FIXME
```

Comment:
A FIXME or TODO should have a corresponding issue and the issue number should be included here.

Reply:
Done